src/main/java/org/terasology/fluid/system/FluidAuthoritySystem.java

```diff
@@ -101,63 +117,104 @@ public void initialise() {
-        Vector3f location = new Vector3f(start);
-        Vector3f normalizedDirection = new Vector3f(direction).normalize();
-        for (int i = 0; i < distance; i++) {
-            location.add(normalizedDirection);
-            Vector3i intLocation = new Vector3i(location, RoundingMode.HALF_DOWN);
-            Block block = worldProvider.getBlock(intLocation);
-            if (block.isLiquid()) {
-                return Optional.of(intLocation);
-            }
-            if (!block.isPenetrable()) {
-                return Optional.empty();
-            }
+    private Optional<Vector3i> getLiquidInReach(final Vector3f start, final Vector3f direction, EntityRef character, float distance) {
+        HitResult hitResult = physics.rayTrace(start, direction, distance, Sets.newHashSet(character), PHYSICSFILTER);
+        if (!hitResult.isHit() || !hitResult.isWorldHit()) {
+            return Optional.empty();
+        }
+        Vector3i pos = hitResult.getBlockPosition();
+        if (worldProvider.getBlock(pos).isLiquid()) {
+            return Optional.of(pos);
+        } else {
+            return Optional.empty();
+        }
+    }
+    private Optional<Vector3i> getPlacementPosition(final Vector3f start, final Vector3f direction, EntityRef character, float distance) {
+        HitResult hitResult = physics.rayTrace(start, direction, distance, Sets.newHashSet(character), PHYSICSFILTER);
+        if (!hitResult.isHit() || !hitResult.isWorldHit()) {
+            return Optional.empty();
+        }
+        Vector3i pos = new Vector3i(hitResult.getBlockPosition());
+        pos.add(Side.inDirection(hitResult.getHitNormal()).direction());
+        if (worldProvider.getBlock(pos).equals(air)) {
+            return Optional.of(pos);
+        } else {
+            return Optional.empty();
         }
-        return Optional.empty();
     }
 
     /**
      * Fill up the provided fluid container item with the current fluid interacted with in the game world.
-     *
-     * @param event          The event which has details about how this entity was activated.
-     * @param item           The reference to the item being activated.
-     * @param fluidContainer The component used for storing fluid in a container.
-     * @param itemComponent  A component included for filtering out non-matching events. Here, we only want entities
-     *                       which are used as items.
      */
+    @ReceiveEvent(priority = EventPriority.PRIORITY_HIGH)
+    public void fillFluidContainerItem(AttackRequest event, EntityRef character, CharacterComponent characterComponent) {
+        EntityRef item = event.getItem();
+        FluidContainerItemComponent fluidContainer = item.getComponent(FluidContainerItemComponent.class);
+        if (fluidContainer != null && (fluidContainer.fluidType == null || fluidContainer.volume < fluidContainer.maxVolume)) {
```

Comment:
I think these nested levels could be replaced with a condition then an empty return.